### DISCUSSION TITLE
Support GitHub App Authentication

### DISCUSSION THREAD

**Discussion by jamietanna** [renovatebot/renovate#21035] [Suggest an Idea]
### What would you like Renovate to be able to do?

Given a `GITHUB_APP_KEY` and a `GITHUB_APP_ID`, perform [GitHub App Authentication](https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps) for the application, rather than requiring a user pre-set this installation access token.

We would likely need to take into account the token's expiry, as it lasts for one hour, which may require a bit more of a refactor to allow for authentication to be re-fetched if 401/403s are returned.

### If you have any ideas on how this should be implemented, please tell us here.

We currently implement this in Mend Renovate On Prem - of which no source code is available - but I assume it is retrieving a token and then passing it through to Renovate in the `token` parameter, as we're expecting folks to do so [when running as a GitHub App](https://docs.renovatebot.com/modules/platform/github/#running-as-a-github-app).

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by Shegox** [comment]
Also for the GitHub App we need to take into account that the GitHub installation token is scoped only at most to a single organisation. Meaning that to renovate all repositories where an App has access, it needs to perform following calls: 
1. List all Installations
2. Create an access token for each installation (==organisation) and use that to authenticate to GitHub. 
3. List all repos the specific installation has access to.
4. Perform normal renovate procedure.

Alternatively you can as well scope the access token to just a repository (instead of the whole organisation). If we can support that, we need to ideally hook into the process between the renovation of each repository to update/change the token depending on the repository.

**Comment by voidlily** [comment]
This may be possible with renovate-on-prem? https://github.com/mend/renovate-on-prem

**Comment by jamietanna** [comment]
It is, but my end-use-case is things that are built on top of Renovate such as https://gitlab.com/tanna.dev/renovate-graph, as well as anyone else using OSS Renovate

**Comment by viceice** [comment]
this isn't easy to implement without refactoring a lot of renovate internals.

@rarkins i think we declined this before? 🤔

**Comment by JamieMagee** [comment]
Yeah, was declined due to added maintenance burden in #8196.

I think this issue was created out of [a discussion on Slack][1], to see if it would be accepted now.

[1]: https://renovatebot.slack.com/archives/CAFH752JU/p1674163161056719

**Comment by rarkins** [comment]
This could be simplified if:
- Operate on a single org at a time
- Use same token on each repo
- The operation lasts for at most an hour

Because in that case you could just generate an installation token for the org and use the same one for the entire run.

**Comment by jamietanna** [comment]
Thanks folks for the feedback. 

For anyone interested in implementing it on top of Renovate, [these are the changes](https://gitlab.com/tanna.dev/renovate-graph/-/merge_requests/1) I made for renovate-graph.

I take the point that it's going to be a significant rewrite of several portions of Renovate, so may not make sense right now, especially as there's already support in the Mend Renovate On-Prem offering.

**Comment by NCSamiAhonen** [comment]
Hi, This would be needed feature. Today just looked thru this, I have a use case that I want to install renovate pods into Kubernetes cluster and access from there into GitHub by using GitHubApp token creation mechanism, other tokens gives too wide accesses into repositories when using static token approach. Seems that this is not possible atm?

**Comment by rarkins** [comment]
You can use a github app installation token today, it just needs to be generated on the fly before you pass it to Renovate

**Comment by jamietanna** [comment]
Or as mentioned above there's https://github.com/mend/renovate-on-prem which provides Helm charts for easy deployment to Kubernetes, and can use GitHub App auth :)

**Comment by viceice** [comment]
there is also a sample configuration somewhere in our helm chart repo.

**Comment by NCSamiAhonen** [comment]
Ok. Thanks i will look these. 👍

**Comment by knechtionscoding** [comment]
> You can use a github app installation token today, it just needs to be generated on the fly before you pass it to Renovate

Having an example of this/adding it to the helm chart as an init container or side car container would go a long way; having everyone who needs this reimplement it seems excessive if there is a working example somewhere.

Also, is dumping it into an env variable viable? Will renovate read the updated `RENOVATE_TOKEN` env variable on every call to github? Or does it only read it once at boot? If Renovate doesn't currently read the env variable on each call, it might be a reasonable middle ground to make it do so, and then we can have a separate sidecar/process that updates that env variable appropriately? I'd assume that requires less rewrite of the renovate internals?

Edit: I suppose that would involve updating the env variables of a child process (because renovate doesn't run as pid 1) and I have a feeling that's a lot of work.

**Comment by viceice** [comment]
> > You can use a github app installation token today, it just needs to be generated on the fly before you pass it to Renovate
> 
> Having an example of this/adding it to the helm chart as an init container or side car container would go a long way; having everyone who needs this reimplement it seems excessive if there is a working example somewhere.
> 
> Also, is dumping it into an env variable viable? Will renovate read the updated `RENOVATE_TOKEN` env variable on every call to github? Or does it only read it once at boot? If Renovate doesn't currently read the env variable on each call, it might be a reasonable middle ground to make it do so, and then we can have a separate sidecar/process that updates that env variable appropriately? I'd assume that requires less rewrite of the renovate internals?
> 
> Edit: I suppose that would involve updating the env variables of a child process (because renovate doesn't run as pid 1) and I have a feeling that's a lot of work.
> 
> 

won't work. renovate internals don't allow to update the token after initial configuration. too complicated changes would be required .

btw: there is a sample somewhere, probably inside our helm chart repo.

**Reply by knechtionscoding** [reply]
> btw: there is a sample somewhere, probably inside our helm chart repo

I was unable to find that example in the charts repo or the docs. My googling/got search skills might be to blame but is it possible for you to point me at at it?

**Reply by vrivellino** [reply]
FWIW - I was able to implement on-demand generation of an installation token using the helm chart when deployed as a cronjob:

```yaml
cronjob:
  # Run every morning
  schedule: "35 7 * * *"
  enabled: true
  concurrencyPolicy: Forbid
  preCommand: |
    echo "Generating GitHub App JWT..."
    HEADER=$(printf '{"alg":"RS256","typ":"JWT"}' | openssl base64 -A | tr '+/' '-_' | tr -d '=')
    NOW=$(date +%s)
    EXP=$((NOW + 540))
    PAYLOAD=$(printf '{"iat":%d,"exp":%d,"iss":"%s"}' "$NOW" "$EXP" "$RENOVATE_GITHUB_APP_ID" | openssl base64 -A | tr '+/' '-_' | tr -d '=')
    JWT="${HEADER}.${PAYLOAD}"
    SIGNATURE=$(printf '%s' "$JWT" | openssl dgst -sha256 -sign <(printf '%s' "$RENOVATE_GITHUB_APP_KEY") | openssl base64 -A | tr '+/' '-_' | tr -d '=')
    JWT="${JWT}.${SIGNATURE}"

    echo "Requesting installation token..."
    installation_token="$(curl -s -X POST \
      -H "Authorization: Bearer ${JWT}" \
      -H "Accept: application/vnd.github+json" \
      "https://api.github.com/app/installations/${RENOVATE_GITHUB_APP_INSTALLATION_ID}/access_tokens" \
      | jq -r '.token')"
    export RENOVATE_TOKEN="$installation_token"

envFrom:
  - secretRef:
      name: renovate-github-app  # contains RENOVATE_GITHUB_APP_ID, RENOVATE_GITHUB_APP_INSTALLATION_ID, and RENOVATE_GITHUB_APP_KEY
```

This uses the `preCommand` functionality provided by the chart to execute arbitrary shell commands, and uses curl and openssl to  call github to create a temporary installation token.

Installations token expire after an hour, so this workaround is only useful when renovate executes as a cronjob.

**Reply by harish2k01** [reply]
I was searching for a while now, and the preCommand worked as a charm. Thanks!

**Comment by rarkins** [comment]
Converted to a Discussion as this is not a feature we would like to add

**Comment by jamietanna** [comment]
Related: https://github.com/renovatebot/renovate/issues/11325

However, as this is more involved, it's not as likely to be implemented compared to some of the other suggestions
